### PR TITLE
Move SSGTS test scenarios closer to rule definitions

### PR DESCRIPTION
**Pull request by jan-cerny**
#### Description:
The test scenarios for a rule are moved to `tests` subdirectory in the rule directory.

#### Rationale:
Developers who work on a rule will not have to search for the scenarios in a different directory tree.
It is cumbersome to add, or find test scenarios for a rule.

### CODE REVIEW

File: tests/ssg_test_suite/common.py [hunk 1]
```diff
@@ -185,11 +195,35 @@ def run_with_stdout_logging(command, args, log_file):
...
+        return None
+    return tarinfo
+
+
+def create_tarball(tar_dir):
+    """Create a tarball which contains all test scenarios for every rule.
+    Tarball contains directories with the test scenarios. The name of the
+    directories is the same as short rule ID. There is no tree structure.
+    """
+    archive_path = os.path.join(tar_dir, _TAR_BASENAME)
```
Comment (jan-cerny) [inline tests/ssg_test_suite/common.py]: I suggest creating the tarball in a temp file.

**Comment by jan-cerny** [comment]
I have updated the PR. A review will be appreciated.

The script that has been used to mass moving is https://paste.fedoraproject.org/paste/FLZ1LyPnj~o5vMpqxQijeA

Please don't forget that the commit also changes the Python code in test suite.

**Comment by matejak** [comment]
Could you please separate the test suite code changes and scenario relocations to two commits? It would make this PR slightly easier to review.

**Comment by jan-cerny** [comment]
@matejak yes, I could, if you don't mind that the commits will be destructive if used separately

**Comment by jan-cerny** [comment]
@matejak I have split it for you.

File: tests/ssg_test_suite/rule.py [hunk 2]
```diff
@@ -187,13 +186,13 @@ def _final_scan_went_ok(self, runner, rule_id):
             logging.error(msg)
         return success
 
-    def _matches_target(self, rule_dir, targets):
```
Comment (matejak) [inline tests/ssg_test_suite/rule.py]: Nice naming improvement.

File: tests/ssg_test_suite/common.py [hunk 3]
```diff
@@ -220,5 +261,31 @@ def send_scripts(domain_ip):
...
+        if "rule.yml" in filenames and "tests" in dirnames:
+            short_rule_id = os.path.basename(dirpath)
+            tests_dir = os.path.join(dirpath, "tests")
+            tests_dir_files = os.listdir(tests_dir)
+            # Filter out everything except the shell test scenarios.
+            # Other files in rule directories are editor swap files
+            # or other content than a test case.
+            scripts = filter(lambda x: x.endswith(".sh"), tests_dir_files)
+            full_rule_id = OSCAP_RULE + short_rule_id
+            result = Rule(tests_dir, full_rule_id, short_rule_id, scripts)
```
Comment (matejak) [inline tests/ssg_test_suite/common.py]: Please use the kwargs here, so it is obvious that the `result.id == full_rule_id` and the assignment is robust wrt to the `Rule` type possible evolution.

File: tests/ssg_test_suite/common.py [hunk 4]
```diff
@@ -220,5 +261,31 @@ def send_scripts(domain_ip):
...
+
+def iterate_over_rules():
+    """Iterate over rule directories which have test scnearios".
+
+    Returns:
+        Named tuple Rule having these fields:
+            directory -- absolute path to the rule "tests" subdirectory
+                         containing the test scenarios in Bash
+            id -- full rule id as it is present in datastream
+            short_id -- short rule ID, the same as basename of the
```
Comment (matejak) [inline tests/ssg_test_suite/common.py]: basename of the ... directory -> please shuffle those lines.

File: tests/README.md [hunk 5]
```diff
@@ -400,8 +366,8 @@ echo "KerberosAuthentication yes" >> /etc/ssh/sshd_config
 
 Let's add test scenarios for rule `accounts_password_minlen_login_defs`
 
-1. Create appropriate directory within `data/` directory tree (in this case
-  `data/group_system/group_accounts/group_accounts-restrictions/group_password_expiration/rule_accounts_password_minlen_login_defs` further referenced as *DIR*).
+1. Create `tests` directory within rule directory (in this case
+  `/linux_os/guide/system/accounts/accounts-restrictions/password_expiration/accounts_password_minlen_login_defs/tests` further referenced as *DIR*).
 1. write a few fail scripts - for example removing the line, commenting it, wrong value, etc.
  into *DIR*
```
Comment (matejak) [inline tests/README.md]: Please add a mention of how to use the shared bash code.

**Comment by jan-cerny** [comment]
@matejak I have documented /tests/shared, improved doctext for iterate_over_rules and used named parameters for class Rule. When you will request a rebase ping me.

**Comment by matejak** [comment]
Looks great, go ahead with the rebase!

File: tests/ssg_test_suite/rule.py (hunk 2 above)
Comment (jan-cerny) [reply tests/ssg_test_suite/rule.py]: I don't like when you call a rule a target because it is confused with a target system.

**Comment by jan-cerny** [comment]
@matejak I have rebased it.

**Comment by matejak** [comment]
Great, merging!

**Pull request merged**